### PR TITLE
Introduce inheritance

### PR DESCRIPTION
**Pull request by mcg-web**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Documented?   | yes
| Fixed tickets | none
| License       | MIT

### CODE REVIEW

**Comment by renatomefi** [comment]
I started checking the code, give me some time since it's quite a big change!

**Comment by renatomefi** [comment]
@mcg-web have you seen how gitlab makes their extends?
I think it's a pretty nice syntax:
https://docs.gitlab.com/ee/ci/yaml/#anchors

**Comment by mcg-web** [comment]
@renatomefi this is yaml native syntax but this need to be done in a single file. Inheritance is also a first step to graphql schema language configuration integration. I have also finish it implementation but it's dependent to this PR so I can't submit it right now.

**Comment by mcg-web** [comment]
[see it here](https://github.com/mcg-web/GraphQLBundle/compare/type_inheritance...mcg-web:graphql-schema-language-configuration)

File: Resources/doc/definitions/type-inheritance.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
+Inheritance
+===========
+
+In some cases, inheritance can helps to don't repeat your self, beater organisation,
```
Comment (renatomefi) [inline Resources/doc/definitions/type-inheritance.md]:
```diff
- helps to don't repeat your self
+ help to not repeating yourself
- beater
+ better
```

File: Resources/doc/definitions/type-inheritance.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    config:
+        interfaces: [InterfaceA]
+        fields:
+             foo: {type: String!}
+
+# ObjectA inherited config from InterfaceA and ObjectC
+ObjectA:
+    type: object
+    # ObjectB inherited config (fields, args...) from ObjectA
+    hiers: [ObjectB]
```
Comment (renatomefi) [inline Resources/doc/definitions/type-inheritance.md]: I think you meant heirs? Like the person who inherits? If so I think this typo is system wide

File: Resources/doc/definitions/type-inheritance.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    type: interface
+    config:
+        resolveType: '@=...'
+        fields:
+            bar: {type: String!}
+            bar2: {type: String!}
+            bar3: {type: String!}
+```
+
+This equivalent to:
```
Comment (renatomefi) [inline Resources/doc/definitions/type-inheritance.md]:
```diff
- This 
+ This is
```

File: Resources/doc/definitions/type-inheritance.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    config:
+        resolveType: '@=...'
+        fields:
+            bar: {type: String!}
+            bar2: {type: String!}
+            bar3: {type: String!}
+```
+
+**Notes:**
+ * Inheritance works only with the configuration right now
```
Comment (renatomefi) [inline Resources/doc/definitions/type-inheritance.md]: maybe?
```diff
- with the
+ with the yaml
```

File: Resources/doc/definitions/type-inheritance.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+ * Type can inherited from any type (from different files)
```
Comment (renatomefi) [inline Resources/doc/definitions/type-inheritance.md]:
```diff
- can
+ can be
```

Comment (renatomefi) [reply Resources/doc/definitions/type-inheritance.md]:
```diff
- from
+ even from
```

File: Resources/doc/definitions/type-inheritance.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+ * Inheritance priority is define by the order defined in `extends` section.
```
Comment (renatomefi) [inline Resources/doc/definitions/type-inheritance.md]:
```diff
-  priority is define
+  priority is defined
- order defined
+ order
- in extends
+ in the extends
```

File: Resources/doc/definitions/type-inheritance.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+ * Type must be of same type to be extended, only `object` types can also extend
+   `interface` type
+ * Inheritance use internally `array_replace_recursive` php function.
+   for example ObjectA config is the result
```
Comment (renatomefi) [inline Resources/doc/definitions/type-inheritance.md]:
```diff
-  the result
+  the result of
```

File: Resources/doc/definitions/type-inheritance.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+   `array_replace_recursive(ObjectCConfig, InterfaceAConfig, ObjectAConfig)`
+
+You can also create decorator types to be use as reusable templates
```
Comment (renatomefi) [inline Resources/doc/definitions/type-inheritance.md]:
```diff
-  use
+  used
```

File: Resources/doc/definitions/type-inheritance.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+without creating useless types and will not be effectively generate
```
Comment (renatomefi) [inline Resources/doc/definitions/type-inheritance.md]: I think this is not so clear, maybe you can rephrase it?

File: Resources/doc/definitions/type-inheritance.md [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+(so decorator type should not be reference as type in fields).
+
+Here an example of a decorator type:
```
Comment (renatomefi) [inline Resources/doc/definitions/type-inheritance.md]:
```diff
-  here
+  here is
```

File: Config/TypeWithOutputFieldsDefinition.php [hunk 11]
```diff
@@ -2,119 +2,15 @@
...
-use Overblog\GraphQLBundle\Relay\Mutation\MutationFieldDefinition;
-use Overblog\GraphQLBundle\Relay\Node\GlobalIdFieldDefinition;
-use Overblog\GraphQLBundle\Relay\Node\NodeFieldDefinition;
-use Overblog\GraphQLBundle\Relay\Node\PluralIdentifyingRootFieldDefinition;
 use Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition;
 use Symfony\Component\Config\Definition\Builder\TreeBuilder;
-use Symfony\Component\Config\Definition\Exception\InvalidConfigurationException;
 
 abstract class TypeWithOutputFieldsDefinition extends TypeDefinition
 {
```
Comment (renatomefi) [inline Config/TypeWithOutputFieldsDefinition.php]: Great cleanup here! :100:

File: DependencyInjection/TypesConfiguration.php [hunk 12]
```diff
@@ -80,6 +75,10 @@ function ($type) {
                 ->cannotBeOverwritten()
                 ->children()
                     ->enumNode('type')->values(self::$types)->isRequired()->end()
+                    ->arrayNode('extends')
+                        ->prototype('scalar')->info('Types to extends.')->end()
```
Comment (renatomefi) [inline DependencyInjection/TypesConfiguration.php]:
```diff
- Types to extends
+ Types to extend
```

File: Resources/doc/definitions/type-inheritance.md (hunk 2 above)
Comment (renatomefi) [reply Resources/doc/definitions/type-inheritance.md]: I think a good word for the configuration file could be `inherits` and internally when doing to inverse reference could indeed be `heirs`

File: Config/Processor/InheritanceProcessor.php [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+    private static function processConfigsHiers(array $configs)
+    {
+        foreach ($configs as $parentName => &$config) {
+            if (!empty($config['hiers'])) {
+                if (!is_array($config['hiers'])) {
+                    $config['hiers'] = [$config['hiers']];
+                }
+                foreach ($config['hiers'] as $hier) {
+                    if (!isset($configs[$hier])) {
+                        throw new \InvalidArgumentException(sprintf(
```
Comment (renatomefi) [inline Config/Processor/InheritanceProcessor.php]: Maybe you could add some tests for this scenario?

File: Config/Processor/InheritanceProcessor.php [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+            return !isset($config['decorator']) || true !== $config['decorator'];
+        });
+    }
+
+    private static function processConfigsHiers(array $configs)
+    {
+        foreach ($configs as $parentName => &$config) {
+            if (!empty($config['hiers'])) {
+                if (!is_array($config['hiers'])) {
+                    $config['hiers'] = [$config['hiers']];
```
Comment (renatomefi) [inline Config/Processor/InheritanceProcessor.php]: when can this happen?
I mean, how the yaml would look like?

File: Config/Processor/BuilderProcessor.php [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+    {
+        self::checkBuilderClass($fieldBuilderClass, 'field');
+        self::$fieldBuilderClassMap[$name] = $fieldBuilderClass;
+    }
+
+    /**
+     * @param string $builderClass
+     * @param string $type
+     */
+    private static function checkBuilderClass($builderClass, $type)
```
Comment (renatomefi) [inline Config/Processor/BuilderProcessor.php]: Would it be good to test all those exceptions?

File: Resources/doc/definitions/type-inheritance.md (hunk 2 above)
Comment (renatomefi) [reply Resources/doc/definitions/type-inheritance.md]: But I think you mean the inverse right? So just `heirs` could be good indeed

File: Config/Processor/BuilderProcessor.php (hunk 15 above)
Comment (mcg-web) [reply Config/Processor/BuilderProcessor.php]: I'll add some Unit tests to cover this :+1:

File: Resources/doc/definitions/type-inheritance.md (hunk 2 above)
Comment (mcg-web) [reply Resources/doc/definitions/type-inheritance.md]: Yes it `heirs` the correct word that is the inverse of `extends` right now.
You mean replace `extends` by `inherits`?

File: Resources/doc/definitions/type-inheritance.md (hunk 4 above)
Comment (mcg-web) [reply Resources/doc/definitions/type-inheritance.md]: Inheritance works only with the generated types right now

File: Generator/TypeGenerator.php [hunk 16]
```diff
@@ -4,31 +4,43 @@
...
     private static $classMapLoaded = false;
 
-    public function __construct($classNamespace, array $skeletonDirs, $cacheDir, callable $defaultResolver, array $configs, $useClassMap = true)
+    public function __construct(
+        $classNamespace,
+        array $skeletonDirs,
+        $cacheDir,
+        callable $defaultResolver,
+        array $configs,
+        $useClassMap = true,
```
[the remark this answers is not in the document]
Comment (mcg-web) [reply Generator/TypeGenerator.php]: done

File: Config/Processor/BuilderProcessor.php (hunk 15 above)
Comment (mcg-web) [reply Config/Processor/BuilderProcessor.php]: done

**Pull request merged**